be/src/runtime/stream_load/stream_load_executor.cpp

```diff
@@ -229,49 +244,55 @@ Status StreamLoadExecutor::commit_txn(StreamLoadContext* ctx) {
 #else
     result = k_stream_load_commit_result;
 #endif
-    // Return if this transaction is committed successful; otherwise, we need try
-    // to rollback this transaction.
     Status status(result.status);
-    if (!status.ok()) {
-        LOG(WARNING) << "commit transaction failed, errmsg=" << status.get_error_msg() << ctx->brief();
-        if (status.code() == TStatusCode::PUBLISH_TIMEOUT) {
-            ctx->need_rollback = false;
-            if (ctx->load_deadline_sec > UnixSeconds()) {
-                //wait for apply finish
-                TGetLoadTxnStatusRequest v_request;
-                TGetLoadTxnStatusResult v_result;
-                set_request_auth(&v_request, ctx->auth);
-                v_request.db = ctx->db;
-                v_request.tbl = ctx->table;
-                v_request.txnId = ctx->txn_id;
-                while (ctx->load_deadline_sec > UnixSeconds()) {
-                    sleep(std::min((int64_t)config::get_txn_status_internal_sec,
-                                   ctx->load_deadline_sec - UnixSeconds()));
-                    auto visiable_st = ThriftRpcHelper::rpc<FrontendServiceClient>(
-                            master_addr.hostname, master_addr.port,
-                            [&v_request, &v_result](FrontendServiceConnection& client) {
-                                client->getLoadTxnStatus(v_result, v_request);
-                            },
-                            config::txn_commit_rpc_timeout_ms);
-                    if (!visiable_st.ok()) {
-                        return status;
-                    } else {
-                        if (v_result.status == TTransactionStatus::VISIBLE) {
-                            return Status::OK();
-                        } else if (v_result.status == TTransactionStatus::COMMITTED) {
-                            continue;
-                        } else {
-                            return status;
-                        }
-                    }
-                }
-            }
-        }
+    if (status.ok()) {
+        return status;
+    } else if (status.code() == TStatusCode::PUBLISH_TIMEOUT) {
+        bool visible = wait_txn_visible_until(auth, request.db, request.tbl, request.txnId, deadline);
+        return visible ? Status::OK() : status;
+    } else {
         return status;
     }
-    // commit success, set need_rollback to false
-    ctx->need_rollback = false;
```

Comment:
I don't see the same code after refactoring.

Reply:
Yes you're right, I forgot to set this variable.